### PR TITLE
refactor(protocol-engine): Add models for the load module command

### PR DESCRIPTION
**Pull request by SyntaxColoring**
# Overview

This PR adds public-facing models for a `loadModule` Protocol Engine command. This is the first step to implementing #8207.

# Changelog

* Add public-facing models for `LoadModuleRequest`, `LoadModuleResult`, and `LoadModule`.
* If you attempt to actually execute a `LoadModule` command through Protocol Engine, it will currently raise a `NotImplementedError`.

# Review requests

Review the command model by either:

* reading the Python code
* running `make -C dev OT_API_FF_enableProtocolEngine=1` and inspecting the run management endpoints at http://localhost:31950/redoc or http://localhost:31950/docs.

Make sure:

* We agree with the basic premise. That is, we have a `loadModule` command that:
    1. takes a string representing the kind of module to load
    2. when executed, makes sure that that module is physically connected
    3. upon success, returns a unique ID representing that physical module
* We agree with naming the field `model`, as opposed to `modelName`, `name`, `loadName`, or something else. `model` is taken from [the current draft of JSON protocol v6](https://github.com/Opentrons/opentrons/blob/0f339f45de238183b2c433e67f839363d5177582/shared-data/protocol/schemas/6.json#L266-L282). Note that this is inconsistent with the `loadPipette` command, which takes a `pipetteName`, not a `pipetteModel`.
* We agree that the above field should take camelCase values like `magneticModuleV1` instead of Python Protocol API values like `magnetic module`. Again, this is taken from [the current draft of JSON protocol v6](https://github.com/Opentrons/opentrons/blob/0f339f45de238183b2c433e67f839363d5177582/shared-data/protocol/schemas/6.json#L266-L282), and again, this is inconsistent with the `loadPipette` command.
* We agree with the "compatible loading" behavior of `model` that I've documented.
* We agree that we're not worried about Thermocycler positioning, for now.

# Risk assessment

No risk to functionality, but this is a fairly important API, so we should make sure it's specced right.

### CODE REVIEW

File: api/src/opentrons/protocol_engine/commands/load_module.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+
+LoadModuleCommandType = Literal["loadModule"]
+
+
+class LoadModuleData(BaseModel):
+    """Data required to load a module."""
+
+    # todo(mm, 2021-11-01): Use an enum instead of a str. shared-data defines the
+    # possible model names.
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_engine/commands/load_module.py:19]: Figuring out what to import from where was hurting my brain. There's a lot of confusing duplication of this kind of thing throughout our codebase, and I didn't want to thoughtlessly compound it. But this should be fairly easy to add later on, once that's figured out.

File: api/src/opentrons/protocol_engine/types.py [hunk 2]
```diff
@@ -27,7 +27,7 @@ class DeckSlotLocation(BaseModel):
 
 
 LabwareLocation = Union[DeckSlotLocation]
-"""Union of all legal labware locations."""
+"""Union of all locations where it's legal to load a labware."""
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_engine/types.py:30]: Unrelated change. Avoiding ambiguity of whether it's the labware or the labware's location that's legal.

File: api/src/opentrons/protocol_engine/commands/load_module.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    model: str = Field(
+        ...,
+        example="magneticModuleV1",
+        description=(
+            "The model name of the module to load."
+            "\n\n"
+            "If a different version of this module is physically connected,"
+            " the load will succeed"
+            " if it's deemed compatible with the version you requested."
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_engine/commands/load_module.py:28]: I haven't actually checked, but I *assume* this is the intent and current behavior of our `loadPipette` command. If so, `loadModule` should do the same thing.

**Review by b-cooper — APPROVED** [review]
Looks good in docs and in code 🔑

**Review by sanni-t — APPROVED** [review]
Code LGTM!
But I have concerns about the V6 schema notation of 'model' or 'modelName' for modules. Modules have a hardware 'model' number that hasn't been (historically) necessarily translated to a module name in the API. For example, temperature module has 5 different hardware models (which are reflected in the serial numbers as `TDV01`, `TDV02`..)  but we don't have different names for them in the API. I don't know if we expose this hardware model number to the user anywhere clearly visible, because if we do, that number would confuse a user.

**Comment by SyntaxColoring** [comment]
> But I have concerns about the V6 schema notation of 'model' or 'modelName' for modules. Modules have a hardware 'model' number that hasn't been (historically) necessarily translated to a module name in the API. For example, temperature module has 5 different hardware models (which are reflected in the serial numbers as `TDV01`, `TDV02`..) but we don't have different names for them in the API. I don't know if we expose this hardware model number to the user anywhere clearly visible, because if we do, that number would confuse a user.

Yeah, agreed.

Off the top of my head, the Python Protocol API [claims](https://docs.opentrons.com/v2/new_protocol_api.html?highlight=model#opentrons.protocol_api.contexts.InstrumentContext.model) to return a string like `p300_single_v1.3` as a *pipette* model. Which I guess is analogous to the `TDV01`/`TDV02` strings that you mentioned, so that makes sense?

But in this new API, we're using "model" to mean something closer to the "load name," like `p300_single` for a pipette. Which is definitely confusing.

Codebase-wide, it's unclear to me what our distinctions are between "name," "load name," "model [name]," "revision," "generation," etc. And it's unclear to me when it's expected for one version to be substitutable for another.
